File: worker/src/components/metadata.py
"""Metadata Component"""
import logging
import datetime
import re

import exiftool
from grpc import RpcError

from src.protos.api_pb2_grpc import APIStub
from src.protos.api_pb2 import MediaItemMetadataRequest  # pylint: disable=no-name-in-module
from src.protos.worker_pb2 import MediaItemComponent, METADATA  # pylint: disable=no-name-in-module
from src.components.component import Component
from src.utils import getval_from_dict


class Metadata(Component):
    """Metadata Component"""

    def __init__(self, api_stub: APIStub) -> None:
        super().__init__(MediaItemComponent.Name(METADATA), api_stub)

    # pylint: disable=too-many-statements,too-many-branches
    async def process(self, mediaitem_user_id: str, mediaitem_id: str, mediaitem_file_path: str, _: dict) -> dict:
        """Process required metadata and generate thumbnail from EXIF data"""
        file_path = f'{mediaitem_file_path}/{mediaitem_id}'

        # extract metadata
        result = {}
        result['userId'] = mediaitem_user_id
        result['id'] = mediaitem_id
        result['status'] = 'PROCESSING'
        result['type'] = 'unknown'
        result['category'] = 'default'
        result['sourcePath'] = file_path
        try:
            with exiftool.ExifToolHelper() as ethelper:
                ethelper.check_execute = False
                metadata = ethelper.get_metadata(file_path)[0]
                logging.debug(f'metadata for user {mediaitem_user_id} mediaitem {mediaitem_id}: {metadata}')
                result['mimeType'] = getval_from_dict(metadata, ['File:MIMEType'])
                result['type'] = 'photo' if result['mimeType'] and 'image' in result['mimeType'] else \
                    'video' if result['mimeType'] and 'video' in result['mimeType'] else 'unknown'
                result['width'] = getval_from_dict(metadata, ['EXIF:SensorWidth', 'EXIF:ImageWidth',
                                                              'EXIF:ExifImageWidth', 'File:ImageWidth',
                                                              'PNG:ImageWidth', 'XMP:ExifImageWidth',
                                                              'QuickTime:ImageWidth', 'QuickTime:SourceImageWidth'],
                                                              return_type='int')
                result['height'] = getval_from_dict(metadata, ['EXIF:SensorHeight', 'EXIF:ImageHeight',
                                                               'EXIF:ExifImageHeight', 'File:ImageHeight',
                                                               'PNG:ImageHeight', 'XMP:ExifImageHeight',
                                                               'QuickTime:ImageHeight', 'QuickTime:SourceImageHeight'],
                                                               return_type='int')
                if (result['height'] is None or result['width'] is None) and 'Composite:ImageSize' in metadata:
                    composite_dims = metadata['Composite:ImageSize'].split(' ')
                    if len(composite_dims) == 2:
                        result['width'] = int(composite_dims[0])
                        result['height'] = int(composite_dims[1])
                creation_time = getval_from_dict(metadata, ['EXIF:DateTimeOriginal', 'EXIF:CreateDate', \
                                            'XMP:CreateDate', 'XMP:DateCreated', 'XMP:DateTimeOriginal', \
                                            'Composite:SubSecCreateDate', 'Composite:SubSecDateTimeOriginal', \
                                            'QuickTime:CreateDate', 'QuickTime:TrackModifyDate', \
                                            'QuickTime:MediaCreateDate', 'QuickTime:CreationDate', \
                                            'EXIF:ModifyDate', 'XMP:ModifyDate', \
                                            'File:FileModifyDate', 'File:FileAccessDate', 'File:FileInodeChangeDate'])
                # work(omkar): handle timezone when "its time" :P
                if creation_time and re.search(r'[\+]\d{2}:\d{2}', creation_time):
                    creation_time = creation_time.rsplit("+", maxsplit=1)[0] if creation_time else None
                elif creation_time and re.search(r'[\-]\d{2}:\d{2}', creation_time):
                    creation_time = creation_time.rsplit("-", maxsplit=1)[0] if creation_time else None
                if creation_time:
                    creation_time = creation_time.replace('- ', '-')
                    creation_time = creation_time.replace(' -', '-')
                    creation_time = creation_time.replace('T', ' ') if 'T' in creation_time else creation_time
                    creation_time = creation_time.replace('Z', '') if 'Z' in creation_time else creation_time
                if creation_time and '-' not in creation_time:
                    result['creationTime'] = datetime.datetime.strptime(creation_time, '%Y:%m:%d %H:%M:%S').replace(
                        tzinfo=datetime.timezone.utc).strftime('%Y-%m-%d %H:%M:%S') if creation_time else None
                elif creation_time:
                    result['creationTime'] = datetime.datetime.strptime(creation_time, '%Y-%m-%d %H:%M:%S').replace(
                        tzinfo=datetime.timezone.utc).strftime('%Y-%m-%d %H:%M:%S') if creation_time else None
                camera_make = getval_from_dict(metadata, ['EXIF:Make', 'XMP:Make', 'QuickTime:Make'])
                result['cameraMake'] = camera_make.strip() if camera_make else None
                camera_model = getval_from_dict(metadata, ['EXIF:Model', 'XMP:Model', 'QuickTime:Model'])
                result['cameraModel'] = camera_model.strip() if camera_model else None
                result['focalLength'] = getval_from_dict(metadata, ['EXIF:FocalLength'])
                result['apertureFNumber'] = getval_from_dict(metadata, ['EXIF:FNumber'])
                result['isoEquivalent'] = getval_from_dict(metadata, ['EXIF:ISO', 'XMP:ISO'])
                result['exposureTime'] = getval_from_dict(metadata, ['EXIF:ExposureTime'])
                result['fps'] = getval_from_dict(metadata, ['QuickTime:VideoFrameRate'], return_type='float')
                result['fps'] = str(round(result['fps'])) if result['fps'] is not None else None
                result['latitude'] = getval_from_dict(metadata, ['Composite:GPSLatitude'], return_type='float')
                result['longitude'] = getval_from_dict(metadata, ['Composite:GPSLongitude'], return_type='float')
                if result['latitude'] is None or result['longitude'] is None:
                    if 'EXIF:GPSLatitudeRef' in metadata and 'EXIF:GPSLatitude' in metadata and \
                        'EXIF:GPSLongitudeRef' in metadata and 'EXIF:GPSLongitude' in metadata and \
                         metadata['EXIF:GPSLatitudeRef'] != '' and metadata['EXIF:GPSLatitude'] != '' and \
                         metadata['EXIF:GPSLongitudeRef'] != '' and metadata['EXIF:GPSLongitude'] != '':
                        result['latitude'] = float(metadata['EXIF:GPSLatitude']) * (1 \
                            if metadata['EXIF:GPSLatitudeRef'] == 'N' else -1)
                        result['longitude'] = float(metadata['EXIF:GPSLongitude']) * (1 \
                            if metadata['EXIF:GPSLongitudeRef'] == 'E' else -1)
                    elif 'Composite:GPSPosition' in metadata:
                        splits = metadata['Composite:GPSPosition'].split()
                        if len(splits) == 2:
                            result['latitude'] = float(splits[0])
                            result['longitude'] = float(splits[1])
                result['category'] = self._get_mediaitem_category(metadata, result)
                logging.debug(f'extracted metadata for user {mediaitem_user_id} mediaitem {mediaitem_id}: {result}')
        except Exception as exp:
            logging.error(
                f'error extracting exif metadata for user {mediaitem_user_id} mediaitem {mediaitem_id}: {str(exp)}')
            result['status'] = 'FAILED'
            self._grpc_save_mediaitem_metadata(result)
            return None

        # metadata, preview and thumbnail extracted, so its READY
        result['status'] = 'READY'
        self._grpc_save_mediaitem_metadata(result)
        logging.info(f'processed metadata for user {mediaitem_user_id} mediaitem {mediaitem_id}')

        return result

    def _grpc_save_mediaitem_metadata(self, result: dict):
        """gRPC call for saving mediaitem metadata"""
        try:
            request = MediaItemMetadataRequest(
                userId=result['userId'],
                id=result['id'],
                status=result['status'],
                mimeType=result['mimeType'] if 'mimeType' in result else None,
                type=result['type'] if 'type' in result else None,
                category=result['category'] if 'category' in result else None,
                width=result['width'] if 'width' in result else None,
                height=result['height'] if 'height' in result else None,
                creationTime=result['creationTime'] if 'creationTime' in result else None,
                cameraMake=result['cameraMake'] if 'cameraMake' in result else None,
                cameraModel=result['cameraModel'] if 'cameraModel' in result else None,
                focalLength=result['focalLength'] if 'focalLength' in result else None,
                apertureFNumber=result['apertureFNumber'] if 'apertureFNumber' in result else None,
                isoEquivalent=result['isoEquivalent'] if 'isoEquivalent' in result else None,
                exposureTime=result['exposureTime'] if 'exposureTime' in result else None,
                fps=result['fps'] if 'fps' in result else None,
                latitude=result['latitude'] if 'latitude' in result else None,
                longitude=result['longitude'] if 'longitude' in result else None,
            )
            _ = self.api_stub.SaveMediaItemMetadata(request)
        except RpcError as rpc_exp:
            logging.error(
                f'error sending metadata for mediaitem {request.id}: {str(rpc_exp)}')

    def _get_mediaitem_category(self, metadata: dict, result: dict) -> str:
        """Get mediaitem category from metadata"""
        if result['type'] == 'photo':
            if 'EXIF:UserComment' in metadata and metadata['EXIF:UserComment'].lower() == 'screenshot' \
                or 'XMP:UserComment' in metadata and metadata['XMP:UserComment'].lower() == 'screenshot':
                return 'screenshot'
            if result['width'] > 10000 and result['height']*4 <= result['width']:
                return 'panorama'
        if result['type'] == 'video':
            if int(result['fps']) > 150:
                return 'slow'
            if 'QuickTime:LivePhotoAuto' in metadata:
                return 'live'
            if 'QuickTime:CaptureMode' in metadata and metadata['QuickTime:CaptureMode'].lower() == 'time-lapse':
                return 'timelapse'
        return 'default'

File: worker/src/metadata.cc
// Copyright 2025 Omkar Prabhu
#include "worker/metadata.h"

#include <grpcpp/grpcpp.h>
#include <spdlog/spdlog.h>
#include <sys/wait.h>

#include <algorithm>
#include <cstdio>
#include <memory>
#include <optional>
#include <regex>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

#include "protos/api.pb.h"
#include "worker/api_client.h"
#include "worker/components.h"

using services::api::APIClient;

namespace components {

namespace metadata {

ExifToolClient::ExifToolClient() {
  if (pipe(in_fd) == -1 || pipe(out_fd) == -1) {
    throw std::runtime_error("pipe failed");
  }

  pid = fork();
  if (pid == -1) {
    throw std::runtime_error("fork failed");
  }

  if (pid == 0) {
    dup2(in_fd[0], STDIN_FILENO);
    dup2(out_fd[1], STDOUT_FILENO);
    close(in_fd[1]);
    close(out_fd[0]);
    execlp(exiftool_path.c_str(), exiftool_path.c_str(), "-stay_open", "True",
           "-@", "-", static_cast<char*>(nullptr));
    _exit(1);
  }

  close(in_fd[0]);
  close(out_fd[1]);

  in_stream = fdopen(in_fd[1], "w");
  out_stream = fdopen(out_fd[0], "r");
  if (!in_stream || !out_stream) {
    throw std::runtime_error("fdopen failed");
  }
}

ExifToolClient::~ExifToolClient() {
  if (in_stream) {
    fprintf(in_stream, "-stay_open\nFalse\n");
    fflush(in_stream);
    fclose(in_stream);
  }
  if (out_stream) {
    fclose(out_stream);
  }
  waitpid(pid, nullptr, 0);
}

std::unordered_map<std::string, std::string> ExifToolClient::Extract(
    const std::string& file_path) {
  std::lock_guard<std::mutex> lock(io_mutex);

  fprintf(in_stream, "-s\n-s\n%s\n-execute\n", file_path.c_str());
  fflush(in_stream);

  std::unordered_map<std::string, std::string> result;
  char buffer[4096];
  while (fgets(buffer, sizeof(buffer), out_stream)) {
    std::string line(buffer);
    if (line.find("{ready}") != std::string::npos) {
      break;
    }
    auto pos = line.find_first_of(':');
    if (pos != std::string::npos) {
      std::string key = line.substr(0, pos);
      std::string value = line.substr(pos + 2, line.length() - pos - 3);
      result[key] = value;
    }
  }

  return result;
}

std::unordered_map<std::string, std::string> Metadata::Extract(
    const std::string& id, const std::string& user_id,
    const std::string& mediaitem_id, const std::string& file_path) {
  std::unordered_map<std::string, std::string> result;
  std::unordered_map<std::string, std::string> exif_data;

  // default values
  MediaItemStatus status = MediaItemStatus::PROCESSING;
  MediaItemType type = MediaItemType::UNKNOWN;
  MediaItemCategory category = MediaItemCategory::DEFAULT;

  MediaItemMetadataRequest request;
  request.set_userid(user_id);
  request.set_mediaitemid(mediaitem_id);

  try {
    exif_data = exif_tool_client_->Extract(file_path);

    SPDLOG_DEBUG("exiftool response: {}", exif_data);

    std::string raw_exifdata = "{";
    for (auto it = exif_data.begin(); it != exif_data.end();) {
      const std::string& key = it->first;
      std::string value = it->second;

      // raw exif data
      raw_exifdata.append("\"");
      raw_exifdata.append(key);
      raw_exifdata.append("\":\"");
      raw_exifdata.append(value);
      raw_exifdata.append("\"");
      if (++it != exif_data.end()) {
        raw_exifdata.append(",");
      }
    }
    raw_exifdata.append("}");
    result["exifdata"] = raw_exifdata;
    request.set_exifdata(result["exifdata"]);

    // other fields
    result["mime_type"] = GetValue(exif_data, {"MIMEType"});
    request.set_mimetype(result["mime_type"]);
    if (result["mime_type"].find("image") != std::string::npos) {
      type = MediaItemType::PHOTO;
    } else if (result["mime_type"].find("video") != std::string::npos) {
      type = MediaItemType::VIDEO;
    }
    result["width"] = GetValue(
        exif_data, {"ExifImageWidth", "ImageWidth", "SourceImageWidth"});
    result["height"] = GetValue(
        exif_data, {"ExifImageHeight", "ImageHeight", "SourceImageHeight"});
    if (result["width"] != "") {
      request.set_width(std::stoi(result["width"]));
    }
    if (result["height"] != "") {
      request.set_height(std::stoi(result["height"]));
    }
    result["creation_time"] =
        GetValue(exif_data, {"CreateDate", "DateCreated", "DateTimeOriginal"});
    std::regex date_re(R"((\d{4}):(\d{2}):(\d{2}) (\d{2}):(\d{2}):(\d{2}))");
    std::smatch date_match;
    if (std::regex_search(result["creation_time"], date_match, date_re)) {
      result["creation_time"] =
          date_match[1].str() + "-" + date_match[2].str() + "-" +
          date_match[3].str() + " " + date_match[4].str() + ":" +
          date_match[5].str() + ":" + date_match[6].str();
    }

    // camera
    result["camera_make"] = GetValue(exif_data, {"Make", "LensMake"});
    request.set_cameramake(result["camera_make"]);

    result["camera_model"] = GetValue(exif_data, {"Model", "LensModel"});
    request.set_cameramodel(result["camera_model"]);

    result["focal_length"] = GetValue(exif_data, {"FocalLength"});
    request.set_focallength(result["focal_length"]);

    result["aperture_fnumber"] = GetValue(exif_data, {"FNumber"});
    request.set_aperturefnumber(result["aperture_fnumber"]);

    result["iso_equivalent"] = GetValue(exif_data, {"ISO"});
    request.set_isoequivalent(result["iso_equivalent"]);

    result["exposure_time"] = GetValue(exif_data, {"ExposureTime"});
    request.set_exposuretime(result["exposure_time"]);

    result["fps"] = GetValue(exif_data, {"VideoFrameRate"});
    request.set_fps(result["fps"]);

    result["megapixels"] = GetValue(exif_data, {"Megapixels"});
    request.set_megapixels(result["megapixels"]);

    // gps
    result["longitude"] = GetValue(exif_data, {"GPSLongitude"});
    if (result["longitude"] != "") {
      result["longitude"] = GetCoordinates(result["longitude"]);
      request.set_longitude(std::stod(result["longitude"]));
    }
    result["latitude"] = GetValue(exif_data, {"GPSLatitude"});
    if (result["latitude"] != "") {
      result["latitude"] = GetCoordinates(result["latitude"]);
      request.set_latitude(std::stod(result["latitude"]));
    }

    // category
    if (exif_data.find("UserComment") != exif_data.end()) {
      std::string value = exif_data["UserComment"];
      std::transform(value.begin(), value.end(), value.begin(),
                     [](unsigned char c) { return std::tolower(c); });
      if (value == "screenshot") {
        category = MediaItemCategory::SCREENSHOT;
      }
    } else if (exif_data.find("LivePhotoVideoIndex") != exif_data.end()) {
      category = MediaItemCategory::LIVE;
    } else if (exif_data.find("FullFrameRatePlaybackIntent") !=
               exif_data.end()) {
      if (exif_data["FullFrameRatePlaybackIntent"] == "0") {
        category = MediaItemCategory::TIMELAPSE;
      }
    } else if (exif_data.find("CaptureMode") != exif_data.end()) {
      std::string value = exif_data["CaptureMode"];
      std::transform(value.begin(), value.end(), value.begin(),
                     [](unsigned char c) { return std::tolower(c); });
      if (value == "time-lapse") {
        category = MediaItemCategory::TIMELAPSE;
      }
    } else if (result["width"] != "" && result["height"] != "") {
      int width = std::stoi(result["width"]);
      int height = std::stoi(result["height"]);
      if (width >= 10000 && (height * 4 <= width)) {
        category = MediaItemCategory::PANORAMA;
      }
    } else if (result["fps"] != "") {
      int fps = std::stoi(result["fps"]);
      if (fps >= 120) {
        category = MediaItemCategory::SLOW;
      }
    }
  } catch (const std::exception& e) {
    SPDLOG_ERROR("error extracting metadata: {}", e.what());
    status = MediaItemStatus::FAILED;
  }

  result["status"] = MediaItemStatus_Name(status);
  result["type"] = MediaItemType_Name(type);
  result["category"] = MediaItemCategory_Name(category);

  request.set_status(status);
  request.set_type(type);
  request.set_category(category);
  api_client_->SaveMediaItemMetadata(request);

  return result;
}

std::string GetCoordinates(const std::string& location) {
  int degrees = 0, minutes = 0;
  double seconds = 0.0;
  char direction = 'N';
  double decimal = 0.0;

  if (std::sscanf(location.c_str(), "%d%*[^0-9]%d%*[^0-9]%lf%*[^NSEW]%c",
                  &degrees, &minutes, &seconds, &direction) == 4) {
    decimal = degrees + minutes / 60.0 + seconds / 3600.0;
    if (direction == 'S' || direction == 'W') {
      decimal = -decimal;
    }
  }
  return std::to_string(decimal);
}

std::string GetValue(const std::unordered_map<std::string, std::string>& data,
                     const std::vector<std::string>& keys) {
  for (const std::string& key : keys) {
    auto it = data.find(key);
    if (it != data.end()) {
      return it->second; // no double lookup
    }
  }
  return "";
}

std::shared_ptr<Metadata> Init(const ComponentConfig& config,
                               std::shared_ptr<APIClient> api_client) {
  return std::make_shared<Metadata>(std::make_shared<ExifToolClient>(),
                                    api_client);
}

} // namespace metadata

} // namespace components
